File: src/ros_apps/carla_telemetry/carla_telemetry/sensors/gps.py
"""
| File: gps.py
| Description: CARLA GPS sensor with realistic noise model (Gauss-Markov + white noise).
               Wraps the CARLA sensor.other.gnss sensor and applies the same
               noise model as Micropolis.Telemetry.sensors.gps.GPS.

               The Gauss-Markov bias drift and white-noise are applied additively
               in metres (ENU), then converted back to lat/lon degrees via a
               spherical-Earth reprojection identical to the Micropolis model.
"""
__all__ = ["CarlaGPS"]

import threading
import time
import weakref

import numpy as np

# ---------------------------------------------------------------------------
# Minimal WGS-84 reprojection helper (identical to Micropolis GPS)
# ---------------------------------------------------------------------------
EARTH_RADIUS = 6_353_000.0  # metres


def _reprojection(position: np.ndarray, origin_lat_rad: float, origin_lon_rad: float):
    """Convert a local ENU offset [x_east, y_north, z_up] (metres) to (lat_rad, lon_rad)."""
    x_rad = position[1] / EARTH_RADIUS   # north component
    y_rad = position[0] / EARTH_RADIUS   # east component
    c = np.sqrt(x_rad * x_rad + y_rad * y_rad)
    sin_c = np.sin(c)
    cos_c = np.cos(c)

    if c > 1e-10:
        lat = np.arcsin(
            cos_c * np.sin(origin_lat_rad)
            + (x_rad * sin_c * np.cos(origin_lat_rad)) / c
        )
        lon = origin_lon_rad + np.arctan2(
            y_rad * sin_c,
            c * np.cos(origin_lat_rad) * cos_c
            - x_rad * np.sin(origin_lat_rad) * sin_c,
        )
    else:
        lat = origin_lat_rad
        lon = origin_lon_rad

    return lat, lon


def _latlon_to_enu_offset(lat_rad: float, lon_rad: float,
                           origin_lat_rad: float, origin_lon_rad: float) -> np.ndarray:
    """Approximate inverse: WGS-84 (lat, lon) → ENU metres from origin."""
    dlat = lat_rad - origin_lat_rad
    dlon = lon_rad - origin_lon_rad
    y_north = dlat * EARTH_RADIUS                                        # North (ENU Y)
    x_east = dlon * EARTH_RADIUS * np.cos(origin_lat_rad)               # East  (ENU X)
    return np.array([x_east, y_north, 0.0])


# ---------------------------------------------------------------------------
# CARLA GPS wrapper
# ---------------------------------------------------------------------------

class CarlaGPS:
    """CARLA GNSS sensor wrapper with Gauss-Markov + white-noise model.

    The noise model is identical to ``Micropolis.Telemetry.sensors.gps.GPS``:
    - A Gauss-Markov random-walk bias is integrated each update step.
    - White noise is drawn independently each step.
    - Noise is added in ENU metres, then reprojected to WGS-84.

    Output state dict keys:
    =================== =============================================
    Key                 Value
    =================== =============================================
    ``latitude``        Noisy latitude in **degrees**
    ``longitude``       Noisy longitude in **degrees**
    ``altitude``        Noisy altitude in **metres** (ASL)
    ``latitude_gt``     Ground-truth latitude (degrees)
    ``longitude_gt``    Ground-truth longitude (degrees)
    ``altitude_gt``     Ground-truth altitude (metres)
    ``speed``           Horizontal speed (m/s)
    ``velocity_north``  North velocity (m/s)
    ``velocity_east``   East velocity (m/s)
    ``velocity_down``   Down velocity (m/s)
    ``fix_type``        3 (fixed GPS fix)
    ``eph``             Horizontal position error estimate (m)
    ``epv``             Vertical   position error estimate (m)
    ``satellites``      Simulated visible satellites count
    =================== =============================================
    """

    def __init__(self, carla_world, vehicle_actor, config: dict,
                 origin_lat: float, origin_lon: float, origin_alt: float):
        """
        Args:
            carla_world:    carla.World instance.
            vehicle_actor:  The ego vehicle actor to attach GNSS to.
            config:         GPS config dict from carla_interface_config.yaml.
            origin_lat/lon/alt: World-origin geographic coordinates.
        """
        self._origin_lat = origin_lat
        self._origin_lon = origin_lon
        self._origin_alt = origin_alt
        self._update_rate = float(config.get("update_rate", 10.0))

        # Noise model parameters (identical to Micropolis)
        self._gps_bias = np.zeros(3)
        self._gps_xy_random_walk = float(config.get("gps_xy_random_walk", 2.0))
        self._gps_z_random_walk = float(config.get("gps_z_random_walk", 4.0))
        self._gps_correlation_time = float(config.get("gps_correlation_time", 60.0))

        self._gps_xy_noise_density = float(config.get("gps_xy_noise_density", 2e-4))
        self._gps_z_noise_density = float(config.get("gps_z_noise_density", 4e-4))
        self._gps_vxy_noise_density = float(config.get("gps_vxy_noise_density", 0.2))
        self._gps_vz_noise_density = float(config.get("gps_vz_noise_density", 0.4))

        self._eph = float(config.get("eph", 1.0))
        self._epv = float(config.get("epv", 1.0))
        self._satellites = int(config.get("satellites_visible", 10))

        self._lock = threading.Lock()
        self._last_gnss = None        # raw carla.GnssMeasurement
        self._last_velocity = None    # carla.Vector3D

        # Current published state
        self._state = {
            "latitude": origin_lat, "longitude": origin_lon, "altitude": origin_alt,
            "latitude_gt": origin_lat, "longitude_gt": origin_lon, "altitude_gt": origin_alt,
            "speed": 0.0, "velocity_north": 0.0, "velocity_east": 0.0, "velocity_down": 0.0,
            "fix_type": 3, "eph": self._eph, "epv": self._epv, "satellites": self._satellites,
        }

        # Spawn the CARLA GNSS sensor
        sp_cfg = config.get("spawn_point", {})
        import carla
        bp = carla_world.get_blueprint_library().find("sensor.other.gnss")

        # Apply CARLA noise attributes if configured
        for attr in ("noise_alt_stddev", "noise_lat_stddev", "noise_lon_stddev",
                     "noise_alt_bias", "noise_lat_bias", "noise_lon_bias"):
            val = config.get(attr)
            if val is not None and bp.has_attribute(attr):
                bp.set_attribute(attr, str(float(val)))

        transform = carla.Transform(
            carla.Location(
                x=float(sp_cfg.get("x", 1.0)),
                y=float(sp_cfg.get("y", 0.0)),
                z=float(sp_cfg.get("z", 2.8)),
            ),
            carla.Rotation(
                roll=float(sp_cfg.get("roll", 0.0)),
                pitch=float(sp_cfg.get("pitch", 0.0)),
                yaw=float(sp_cfg.get("yaw", 0.0)),
            ),
        )
        self._sensor = carla_world.spawn_actor(bp, transform, attach_to=vehicle_actor)

        # Weak-ref callback to avoid circular reference
        weak_self = weakref.ref(self)
        self._sensor.listen(lambda data: CarlaGPS._on_gnss(weak_self, data))

    # ------------------------------------------------------------------
    # CARLA callback (fires on CARLA worker thread)
    # ------------------------------------------------------------------

    @staticmethod
    def _on_gnss(weak_self, data):
        """Store latest raw GNSS measurement (thread-safe)."""
        self = weak_self()
        if self is None:
            return
        with self._lock:
            self._last_gnss = data

    # ------------------------------------------------------------------
    # Public API (called from the ROS 2 timer on the main thread)
    # ------------------------------------------------------------------

    def update(self, vehicle_velocity, dt: float):
        """Compute noisy GPS reading and return state dict.

        Args:
            vehicle_velocity: carla.Vector3D — current vehicle velocity (m/s, CARLA frame).
            dt:               Time elapsed since last call (seconds).

        Returns:
            State dict or ``None`` if rate limit not yet reached.
        """
        with self._lock:
            gnss = self._last_gnss

        if gnss is None:
            return None

        # ── Ground-truth lat/lon/alt from CARLA GNSS ──────────────────
        lat_gt_deg = gnss.latitude
        lon_gt_deg = gnss.longitude
        alt_gt = gnss.altitude + self._origin_alt

        origin_lat_rad = np.radians(self._origin_lat)
        origin_lon_rad = np.radians(self._origin_lon)

        # Convert GT to ENU metres for noise injection
        enu_gt = _latlon_to_enu_offset(
            np.radians(lat_gt_deg), np.radians(lon_gt_deg),
            origin_lat_rad, origin_lon_rad,
        )
        enu_gt[2] = gnss.altitude  # use CARLA altitude directly

        # ── Noise model (identical to Micropolis) ─────────────────────
        # Random-walk bias step
        rw = np.array([
            self._gps_xy_random_walk * np.sqrt(dt) * np.random.randn(),
            self._gps_xy_random_walk * np.sqrt(dt) * np.random.randn(),
            self._gps_z_random_walk  * np.sqrt(dt) * np.random.randn(),
        ])

        # Position white noise
        noise_pos = np.array([
            self._gps_xy_noise_density * np.sqrt(dt) * np.random.randn(),
            self._gps_xy_noise_density * np.sqrt(dt) * np.random.randn(),
            self._gps_z_noise_density  * np.sqrt(dt) * np.random.randn(),
        ])

        # Gauss-Markov bias integration
        self._gps_bias += rw * dt - self._gps_bias / self._gps_correlation_time

        # Noisy ENU position
        enu_noisy = enu_gt + noise_pos + self._gps_bias

        # Reproject to lat/lon
        lat_noisy_rad, lon_noisy_rad = _reprojection(enu_noisy, origin_lat_rad, origin_lon_rad)

        # ── Velocity (CARLA frame: X=forward, Y=right, Z=up → NED) ────
        # CARLA velocity: X=forward(North), Y=right(East), Z=up
        vx = float(vehicle_velocity.x)   # forward ≈ North
        vy = float(vehicle_velocity.y)   # right   ≈ East
        vz = float(vehicle_velocity.z)   # up

        # Add velocity noise
        vxy_noise = self._gps_vxy_noise_density * np.sqrt(dt)
        vz_noise = self._gps_vz_noise_density  * np.sqrt(dt)
        vel_north = vx + vxy_noise * np.random.randn()
        vel_east  = vy + vxy_noise * np.random.randn()
        vel_down  = -vz + vz_noise * np.random.randn()

        speed = float(np.sqrt(vx**2 + vy**2))

        self._state = {
            "latitude":       float(np.degrees(lat_noisy_rad)),
            "longitude":      float(np.degrees(lon_noisy_rad)),
            "altitude":       float(enu_noisy[2] + self._origin_alt),
            "latitude_gt":    lat_gt_deg,
            "longitude_gt":   lon_gt_deg,
            "altitude_gt":    alt_gt,
            "speed":          speed,
            "velocity_north": vel_north,
            "velocity_east":  vel_east,
            "velocity_down":  vel_down,
            "fix_type":       3,
            "eph":            self._eph,
            "epv":            self._epv,
            "satellites":     self._satellites,
        }
        return self._state

    @property
    def state(self):
        return self._state

    def destroy(self):
        """Stop the CARLA GNSS sensor."""
        if self._sensor is not None:
            try:
                self._sensor.stop()
                self._sensor.destroy()
            except Exception:
                pass
            self._sensor = None

File: src/ros_apps/carla_telemetry_cpp/src/sensors/gps.cpp
#include "carla_telemetry/sensors/gps.hpp"

#include <carla/actors/BlueprintLibrary.h>
#include <carla/geom/GeoLocation.h>
#include <carla/geom/Transform.h>

#include <cmath>
#include <tuple>

#include "carla_telemetry/perf_monitor.hpp"

namespace carla_telemetry {

namespace {

constexpr double DEG_TO_RAD = M_PI / 180.0;
constexpr double RAD_TO_DEG = 180.0 / M_PI;

constexpr double WGS84_A = 6378137.0;
constexpr double WGS84_B = 6356752.31424518;
constexpr double WGS84_E2 = 0.00669437999014;
constexpr double WGS84_E_PRIME2 = 0.00673949674228;

// Geodetic to ECEF
void geodetic_to_ecef(double lat_rad, double lon_rad, double alt, double& x,
                      double& y, double& z) {
  double sin_lat = std::sin(lat_rad);
  double cos_lat = std::cos(lat_rad);
  double sin_lon = std::sin(lon_rad);
  double cos_lon = std::cos(lon_rad);

  double n = WGS84_A / std::sqrt(1.0 - WGS84_E2 * sin_lat * sin_lat);
  x = (n + alt) * cos_lat * cos_lon;
  y = (n + alt) * cos_lat * sin_lon;
  z = (n * (1.0 - WGS84_E2) + alt) * sin_lat;
}

// ECEF to Geodetic (Bowring's method)
void ecef_to_geodetic(double x, double y, double z, double& lat_rad,
                      double& lon_rad, double& alt) {
  double p = std::sqrt(x * x + y * y);
  if (p < 1e-10) {
    lat_rad = (z > 0.0) ? (M_PI / 2.0) : (-M_PI / 2.0);
    lon_rad = 0.0;
    alt = std::abs(z) - WGS84_B;
    return;
  }

  double theta = std::atan2(z * WGS84_A, p * WGS84_B);
  double sin_theta = std::sin(theta);
  double cos_theta = std::cos(theta);

  lat_rad = std::atan2(
      z + WGS84_E_PRIME2 * WGS84_B * sin_theta * sin_theta * sin_theta,
      p - WGS84_E2 * WGS84_A * cos_theta * cos_theta * cos_theta);
  lon_rad = std::atan2(y, x);

  double sin_lat = std::sin(lat_rad);
  double cos_lat = std::cos(lat_rad);
  alt = p * cos_lat + z * sin_lat -
        WGS84_A * std::sqrt(1.0 - WGS84_E2 * sin_lat * sin_lat);
}

// ECEF to ENU
void ecef_to_enu(double x, double y, double z, double origin_lat_rad,
                 double origin_lon_rad, double origin_alt, double enu[3]) {
  double x0, y0, z0;
  geodetic_to_ecef(origin_lat_rad, origin_lon_rad, origin_alt, x0, y0, z0);

  double dx = x - x0;
  double dy = y - y0;
  double dz = z - z0;

  double sin_lat = std::sin(origin_lat_rad);
  double cos_lat = std::cos(origin_lat_rad);
  double sin_lon = std::sin(origin_lon_rad);
  double cos_lon = std::cos(origin_lon_rad);

  enu[0] = -sin_lon * dx + cos_lon * dy;
  enu[1] = -sin_lat * cos_lon * dx - sin_lat * sin_lon * dy + cos_lat * dz;
  enu[2] = cos_lat * cos_lon * dx + cos_lat * sin_lon * dy + sin_lat * dz;
}

// ENU to ECEF
void enu_to_ecef(const double enu[3], double origin_lat_rad,
                 double origin_lon_rad, double origin_alt, double& x, double& y,
                 double& z) {
  double x0, y0, z0;
  geodetic_to_ecef(origin_lat_rad, origin_lon_rad, origin_alt, x0, y0, z0);

  double sin_lat = std::sin(origin_lat_rad);
  double cos_lat = std::cos(origin_lat_rad);
  double sin_lon = std::sin(origin_lon_rad);
  double cos_lon = std::cos(origin_lon_rad);

  double dx = -sin_lon * enu[0] - sin_lat * cos_lon * enu[1] +
              cos_lat * cos_lon * enu[2];
  double dy = cos_lon * enu[0] - sin_lat * sin_lon * enu[1] +
              cos_lat * sin_lon * enu[2];
  double dz = cos_lat * enu[1] + sin_lat * enu[2];

  x = x0 + dx;
  y = y0 + dy;
  z = z0 + dz;
}

// Convert local ENU offset [x_east, y_north, z_up] (metres) to (lat_rad,
// lon_rad, alt).
std::tuple<double, double, double> reprojection(const double enu[3],
                                                double origin_lat_rad,
                                                double origin_lon_rad,
                                                double origin_alt) {
  double x, y, z;
  enu_to_ecef(enu, origin_lat_rad, origin_lon_rad, origin_alt, x, y, z);
  double lat, lon, alt;
  ecef_to_geodetic(x, y, z, lat, lon, alt);
  return {lat, lon, alt};
}

// WGS-84 (lat, lon, alt) → ENU metres from origin.
void latlon_to_enu_offset(double lat_rad, double lon_rad, double alt,
                          double origin_lat_rad, double origin_lon_rad,
                          double origin_alt, double out[3]) {
  double x, y, z;
  geodetic_to_ecef(lat_rad, lon_rad, alt, x, y, z);
  ecef_to_enu(x, y, z, origin_lat_rad, origin_lon_rad, origin_alt, out);
}

}  // namespace

CarlaGPS::CarlaGPS(carla::client::World& world,
                   carla::SharedPtr<carla::client::Actor> vehicle,
                   const Config& cfg, double origin_lat, double origin_lon,
                   double origin_alt)
    : origin_lat_(origin_lat),
      origin_lon_(origin_lon),
      origin_alt_(origin_alt),
      map_ref_lat_(0.0),
      map_ref_lon_(0.0),
      map_ref_alt_(0.0),
      cfg_(cfg),
      rng_(std::random_device{}()) {
  // Query CARLA map's built-in geo-reference (from OpenDRIVE header).
  // CARLA's GNSS sensor outputs lat/lon relative to THIS origin,
  // so we need it to correctly reproject to user's global_coordinates.
  auto map = world.GetMap();
  if (map) {
    auto geo_ref = map->GetGeoReference();
    map_ref_lat_ = geo_ref.latitude;
    map_ref_lon_ = geo_ref.longitude;
    map_ref_alt_ = geo_ref.altitude;
  }

  // Initialize state
  state_.latitude = origin_lat;
  state_.longitude = origin_lon;
  state_.altitude = origin_alt;
  state_.latitude_gt = origin_lat;
  state_.longitude_gt = origin_lon;
  state_.altitude_gt = origin_alt;
  state_.fix_type = 3;
  state_.eph = cfg.eph;
  state_.epv = cfg.epv;
  state_.satellites = cfg.satellites_visible;

  // Spawn CARLA GNSS sensor
  auto bp_lib = world.GetBlueprintLibrary();
  auto bp = *bp_lib->Find("sensor.other.gnss");

  carla::geom::Transform transform(
      carla::geom::Location(cfg.sp_x, cfg.sp_y, cfg.sp_z),
      carla::geom::Rotation(cfg.sp_pitch, cfg.sp_yaw, cfg.sp_roll));

  auto actor = world.SpawnActor(bp, transform, vehicle.get());
  sensor_ = boost::dynamic_pointer_cast<carla::client::Sensor>(actor);
  sensor_->Listen([this](auto data) { this->on_gnss(data); });
}

void CarlaGPS::on_gnss(carla::SharedPtr<carla::sensor::SensorData> data) {
  auto gnss =
      boost::dynamic_pointer_cast<carla::sensor::data::GnssMeasurement>(data);
  if (!gnss) return;
  // Rate the server delivers GNSS measurements to this client (inter-arrival).
  perf.record_interval("gps.server_dt");
  std::lock_guard<std::mutex> lk(lock_);
  last_gnss_ = gnss;
}

std::optional<GpsState> CarlaGPS::update(double vx, double vy, double vz,
                                         double dt) {
  carla::SharedPtr<carla::sensor::data::GnssMeasurement> gnss;
  {
    std::lock_guard<std::mutex> lk(lock_);
    gnss = last_gnss_;
  }
  if (!gnss) return std::nullopt;

  // CARLA GNSS outputs absolute lat/lon using the map's built-in geo-reference.
  // We need to: (1) convert GNSS → ENU relative to map's geo-ref origin,
  //             (2) then reproject those ENU offsets from user's
  //             global_coordinates.
  double lat_gnss_deg = gnss->GetLatitude();
  double lon_gnss_deg = gnss->GetLongitude();
  double alt_gnss = gnss->GetAltitude();

  double map_ref_lat_rad = map_ref_lat_ * DEG_TO_RAD;
  double map_ref_lon_rad = map_ref_lon_ * DEG_TO_RAD;
  double origin_lat_rad = origin_lat_ * DEG_TO_RAD;
  double origin_lon_rad = origin_lon_ * DEG_TO_RAD;

  // Step 1: Convert GNSS (map-referenced) → ENU metres relative to map origin
  double enu_gt[3];
  latlon_to_enu_offset(lat_gnss_deg * DEG_TO_RAD, lon_gnss_deg * DEG_TO_RAD,
                       alt_gnss, map_ref_lat_rad, map_ref_lon_rad, map_ref_alt_,
                       enu_gt);

  // Ground-truth: reproject ENU from user's global_coordinates origin
  auto [lat_gt_rad, lon_gt_rad, alt_gt] =
      reprojection(enu_gt, origin_lat_rad, origin_lon_rad, origin_alt_);
  double lat_gt_deg = lat_gt_rad * RAD_TO_DEG;
  double lon_gt_deg = lon_gt_rad * RAD_TO_DEG;

  // Noise model (identical to Micropolis)
  double sqrt_dt = std::sqrt(dt);

  // Random-walk bias step
  double rw[3] = {
      cfg_.gps_xy_random_walk * sqrt_dt * normal_(rng_),
      cfg_.gps_xy_random_walk * sqrt_dt * normal_(rng_),
      cfg_.gps_z_random_walk * sqrt_dt * normal_(rng_),
  };

  // Position white noise (using discrete standard deviation, so no sqrt_dt
  // scaling)
  double noise_pos[3] = {
      cfg_.gps_xy_noise_density * normal_(rng_),
      cfg_.gps_xy_noise_density * normal_(rng_),
      cfg_.gps_z_noise_density * normal_(rng_),
  };

  // Gauss-Markov bias integration: x_{k+1} = x_k + rw - x_k * (dt / tau)
  for (int i = 0; i < 3; ++i) {
    gps_bias_[i] += rw[i] - gps_bias_[i] * dt / cfg_.gps_correlation_time;
  }

  // Noisy ENU position
  double enu_noisy[3] = {
      enu_gt[0] + noise_pos[0] + gps_bias_[0],
      enu_gt[1] + noise_pos[1] + gps_bias_[1],
      enu_gt[2] + noise_pos[2] + gps_bias_[2],
  };

  // Cache both ENU vectors so get_latest_enu() can serve them without
  // recomputation, together with the identity of the fix they came from so
  // enu_source() can report it. Written under lock_ because the odometry loop
  // reads the pair (position, source frame) from another thread and must never
  // see one fix's frame number next to another fix's position.
  {
    std::lock_guard<std::mutex> lk(lock_);
    enu_gt_[0] = enu_gt[0];
    enu_gt_[1] = enu_gt[1];
    enu_gt_[2] = enu_gt[2];
    enu_noisy_[0] = enu_noisy[0];
    enu_noisy_[1] = enu_noisy[1];
    enu_noisy_[2] = enu_noisy[2];
    enu_frame_ = static_cast<uint64_t>(gnss->GetFrame());
    enu_sim_time_ = gnss->GetTimestamp();
    enu_valid_ = true;
  }

  // Step 2: Reproject noisy ENU from user's global_coordinates origin
  auto [lat_noisy_rad, lon_noisy_rad, alt_noisy] =
      reprojection(enu_noisy, origin_lat_rad, origin_lon_rad, origin_alt_);

  // Velocity (CARLA frame: X=East, Y=South, Z=up)
  // Therefore: North = -South = -vy, East = vx, Down = -vz
  double vxy_noise = cfg_.gps_vxy_noise_density;  // discrete stddev, no sqrt_dt
  double vz_noise = cfg_.gps_vz_noise_density;
  double vel_north = -vy + vxy_noise * normal_(rng_);
  double vel_east = vx + vxy_noise * normal_(rng_);
  double vel_down = -vz + vz_noise * normal_(rng_);
  double speed = std::sqrt(vx * vx + vy * vy);

  state_.latitude = lat_noisy_rad * RAD_TO_DEG;
  state_.longitude = lon_noisy_rad * RAD_TO_DEG;
  state_.altitude = alt_noisy;
  state_.latitude_gt = lat_gt_deg;
  state_.longitude_gt = lon_gt_deg;
  state_.altitude_gt = alt_gt;
  state_.speed = speed;
  state_.velocity_north = vel_north;
  state_.velocity_east = vel_east;
  state_.velocity_down = vel_down;
  state_.fix_type = 3;
  state_.eph = cfg_.eph;
  state_.epv = cfg_.epv;
  state_.satellites = cfg_.satellites_visible;

  return state_;
}

void CarlaGPS::destroy() {
  if (sensor_) {
    try {
      sensor_->Stop();
      sensor_->Destroy();
    } catch (...) {
    }
    sensor_ = nullptr;
  }
}

bool CarlaGPS::get_latest_gt(double& lat, double& lon, double& alt) {
  std::lock_guard<std::mutex> lk(lock_);
  if (!last_gnss_) return false;

  // Convert GNSS (map-referenced) → ENU → reproject from user's origin
  double map_ref_lat_rad = map_ref_lat_ * DEG_TO_RAD;
  double map_ref_lon_rad = map_ref_lon_ * DEG_TO_RAD;
  double origin_lat_rad = origin_lat_ * DEG_TO_RAD;
  double origin_lon_rad = origin_lon_ * DEG_TO_RAD;

  double enu[3];
  latlon_to_enu_offset(last_gnss_->GetLatitude() * DEG_TO_RAD,
                       last_gnss_->GetLongitude() * DEG_TO_RAD,
                       last_gnss_->GetAltitude(), map_ref_lat_rad,
                       map_ref_lon_rad, map_ref_alt_, enu);

  auto [lat_rad, lon_rad, alt_gt] =
      reprojection(enu, origin_lat_rad, origin_lon_rad, origin_alt_);
  lat = lat_rad * RAD_TO_DEG;
  lon = lon_rad * RAD_TO_DEG;
  alt = alt_gt;
  return true;
}

bool CarlaGPS::get_latest_enu(double& east, double& north, double& up,
                              bool use_noise) const {
  std::lock_guard<std::mutex> lk(lock_);
  if (!enu_valid_) return false;
  const double* src = use_noise ? enu_noisy_ : enu_gt_;
  east = src[0];
  north = src[1];
  up = src[2];
  return true;
}

bool CarlaGPS::enu_source(uint64_t& frame, double& sim_time) const {
  std::lock_guard<std::mutex> lk(lock_);
  if (!enu_valid_) return false;
  frame = enu_frame_;
  sim_time = enu_sim_time_;
  return true;
}

}  // namespace carla_telemetry
